project/Util.scala

```diff
@@ -0,0 +1,60 @@
+import sbt._
+import sbt.Keys._
+import java.nio.file.Files
+import java.util.regex.Pattern
+import scala.collection.JavaConverters._
+
+object Util {
+  lazy val copySharedResources = Def.task {
+//    Called when we have files under core, streaming that need to be copyied into services/js.
+//    ConsoleLogger().info("copying shared resources")
+//    val sharedMessages = "src/main/scala/org/apache/gearpump/shared/Messages.scala"
+//    IO.copyFile(core.base / sharedMessages, file(Build.servicesjs.baseDirectory) / sharedMessages)
+  }
+
+  def findDirs(currentDir: String): Array[File] = {
+    val dirs = IO.listFiles(file(currentDir),  DirectoryFilter)
+    dirs ++ dirs.flatMap(dir => {
+      val childDir = currentDir + "/" + dir.getName
+      findDirs(file(childDir).getPath)
+    })
+  }
+
+  def findFiles(dir: File, pattern: PatternFilter): Array[File] = {
+    val dirs = findDirs(dir.getPath)
+    IO.listFiles(dir, pattern) ++ dirs.flatMap(dir => {
+      IO.listFiles(dir, pattern)
+    })
+  }
+
+  def copyJSArtifactsToOutput: Unit = {
+    ConsoleLogger().info(s"copying JS artifacts ${Build.servicesjs.base.getPath}")
+    val in = file(s"${Build.servicesjs.base.getPath}/target/${Build.scalaVersionMajor}/gearpump-services-js-fastopt.js.map")
+    val out = file(s"${Build.distDashboardDirectory}/gearpump-services-js-fastopt.js.map")
+    val input = IO.read(in)
+    val data = input.replaceAll("../../src","src")
```

Comment:
Ok, this will likely break ~fastOptJS, so files after editing must be manually copied to output. 



> On Aug 3, 2015, at 8:24 PM, Sean Zhong <notifications@github.com> wrote:
> 
> In project/Util.scala:
> 
> > +    })
> > +  }
> > +
> > +  def findFiles(dir: File, pattern: PatternFilter): Array[File] = {
> > +    val dirs = findDirs(dir.getPath)
> > +    IO.listFiles(dir, pattern) ++ dirs.flatMap(dir => {
> > +      IO.listFiles(dir, pattern)
> > +    })
> > +  }
> > +
> > +  def copyJSArtifactsToOutput: Unit = {
> > +    ConsoleLogger().info(s"copying JS artifacts ${Build.servicesjs.base.getPath}")
> > +    val in = file(s"${Build.servicesjs.base.getPath}/target/${Build.scalaVersionMajor}/gearpump-services-js-fastopt.js.map")
> > +    val out = file(s"${Build.distDashboardDirectory}/gearpump-services-js-fastopt.js.map")
> > +    val input = IO.read(in)
> > +    val data = input.replaceAll("../../src","src")
> Here is my proposal:
> 1. remove this function copyJSArtifactsToOutput.
> 2. In sbt/pack target, copy gearpump-services-js-fastopt.js and gearpump-services-js-fastopt.js.map to target\pack\dashboard\views\scalajs\
> 3. If developer want to debug, they can copy services/js/src/ to target/pack/dasboard/
> 
> Here is my reasoning:
> 1. no scala source file sould exist in final package
> 2. Maintianing stuff like findDirs, findFile, or copyJSArtifactsToOutput will complicate the build script.
> 
> —
> Reply to this email directly or view it on GitHub.
>